### PR TITLE
Add core::iter::once_with()

### PR DESCRIPTION
**Pull request by ghost**
Functions `iter::once()` and `iter::repeat()` construct iterators from values. The latter has the lazy variant `iter::repeat_with()`, but the former doesn't. This PR therefore adds `iter::once_with()`.

Another way to think of `iter::once_with()` is that it's a function that converts `FnOnce() -> T` into `Iterator<Item = T>`.

If this seems like a reasonable addition, I'll open a tracking issue and update the `#[feature(...)]` attributes.

### CODE REVIEW

**Comment by frewsxcv** [comment]
cc @rust-lang/libs

**Comment by Centril** [comment]
r? @SimonSapin 

I once promised @eddyb to make a PR for this but the playground ate my work... :(
Thank you for writing this up!

File: src/libcore/iter/sources.rs [hunk 1]
```diff
@@ -377,6 +377,117 @@ pub fn once<T>(value: T) -> Once<T> {
...
+///
+/// // just one, that's all we get
+/// assert_eq!(None, one.next());
+/// ```
+///
+/// Chaining together with another iterator. Let's say that we want to iterate
+/// over each file of the `.foo` directory, but also a configuration file,
+/// `.foorc`:
+///
+/// ```no_run
```
Comment (SimonSapin) [inline src/libcore/iter/sources.rs]: I’d guess here too? IIRC `no_run` doc-tests are still compiled.

File: src/libcore/iter/sources.rs [hunk 2]
```diff
@@ -377,6 +377,119 @@ pub fn once<T>(value: T) -> Once<T> {
     Once { inner: Some(value).into_iter() }
 }
 
+/// An iterator that repeats elements of type `A` endlessly by
+/// applying the provided closure `F: FnMut() -> A`.
+///
```
Comment (kumbayo) [inline src/libcore/iter/sources.rs]: It seems like his comment was copy pasted from `RepeatWith`.
It talks about endless repeating and `FnMut()`.

Comment (Centril) [reply src/libcore/iter/sources.rs]: @kumbayo could you file a PR to fix this?

**Comment by Lucretiel** [comment]
Doesn't `Option` fill this use case, since it implements `IntoIterator`?

**Pull request merged**